# HG changeset patch
# User Rik <rik@octave.org>
# Date 1413571064 25200
#      Fri Oct 17 11:37:44 2014 -0700
# Node ID f8552bd351800aecf190acfcc646e939528a6982
# Parent  a4e993343e9392c7d1793180e49cb68243aed662
unpack.m: Rewrite BIST test to not produce warning on stderr.

* unpack.m: Rewrite BIST test to not produce warning on stderr.

diff --git a/scripts/miscellaneous/unpack.m b/scripts/miscellaneous/unpack.m
--- a/scripts/miscellaneous/unpack.m
+++ b/scripts/miscellaneous/unpack.m
@@ -286,41 +286,47 @@ function files = __parse_bzip2__ (output
   ## Parse the output from bzip2 and bunzip2 returning the files
   ## commpressed (or decompressed).
 
   ## Strip leading blanks and .bz2 extension from file name
   files = regexprep (output, '^\s+(.*)\.bz2: .*', '$1');
 endfunction
 
 
-## FIXME: Maybe there should be a combined test script to test all of the
-##        various pack/unpack routines in the tests/ directory.
 %!test
-%! ## Create temporary file for packing and unpacking
-%! fname = tempname ();
-%! fid = fopen (fname, "wt");
-%! if (fid < 0)
-%!   error ("unpack: Unable to open temporary file for testing");
-%! endif
+%! ## Create temporary directory and file for packing and unpacking
+%! dirname = tempname ();
+%! assert (mkdir (dirname));
+%! filename = tempname ();
+%! fid = fopen (filename, "wt");
+%! assert (fid >= 0);
 %! fprintf (fid, "Hello World\n");
 %! fprintf (fid, "123 456 789\n");
 %! fclose (fid);
-%!
 %! unwind_protect
-%!   copyfile (fname, [fname ".orig"]);
-%!   gzip (fname, P_tmpdir);
-%!   filelist = unpack ([fname ".gz"], P_tmpdir);
-%!   assert (filelist{1}, fname); 
-%!   r = system (sprintf ("diff %s %s.orig", fname, fname));
-%!   if (r)
+%!   copyfile (filename, [filename ".orig"]);
+%!   gzip (filename, dirname);
+%!   [~, f] = fileparts (filename);
+%!   filelist = unpack (fullfile (dirname, [f ".gz"]), P_tmpdir);
+%!   assert (filelist{1}, filename); 
+%!   fid = fopen ([filename ".orig"], "rb");
+%!   assert (fid >= 0);
+%!   orig_data = fread (fid);
+%!   fclose (fid);
+%!   fid = fopen (filename, "rb");
+%!   assert (fid >= 0);
+%!   new_data = fread (fid);
+%!   fclose (fid);
+%!   if (orig_data != new_data)
 %!     error ("unpack: Unpacked file does not equal original");
 %!   endif
 %! unwind_protect_cleanup
-%!   exist (fname) && unlink (fname);
-%!   unlink ([fname ".orig"]);
+%!   unlink (filename);
+%!   unlink ([filename ".orig"]);
+%!   rmdir (dirname);
 %! end_unwind_protect
 
 ## Test input validation
 %!error unpack ()
 %!error unpack (1,2,3,4)
 %!error <FILE must be a string or cell array of strings> unpack (1)
 %!error <file "_%NOT_A_FILENAME%_" not found> unpack ("_%NOT_A_FILENAME%_")
 %!error <file "_%NOT_A_FILENAME%_" not found> unpack ({"_%NOT_A_FILENAME%_"})
